src/bidiMapper/domains/context/logEntry.ts

```diff
@@ -0,0 +1,276 @@
+        '": ' +
+        this._getSingleRemoteValueText((arg.value as any[])[0][1]) +
+        '}'
+      );
+    }
+
+    throw Error('Invalid value type: ' + arg.toString());
+  }
+
+  private _getRemoteValuesText(
+    args: CommonDataTypes.RemoteValue[],
+    formatText: boolean
+  ): string {
+    // if args[0] is a format specifier, format the args as output
+    if (
+      args[0].type === 'string' &&
+      this._isFormmatSpecifier(args[0].value.toString()) &&
+      formatText
+    ) {
+      return this._formatter(args);
+    }
+
+    // if args[0] is not a format specifier, just join the args with \u0020
+    return args
+      .map((arg) => (arg as { type: string; value: any }).value.toString())
+      .join('\u0020');
+  }
+
+  // convert CDP StackTrace object to Bidi StackTrace object
+  private _getBidiStackTrace(
+    cdpStackTrace: Protocol.Runtime.StackTrace | undefined
+  ): Script.StackTrace | undefined {
+    const stackFrames = cdpStackTrace?.callFrames.map((callFrame) => {
+      return {
+        columnNumber: callFrame.columnNumber,
+        functionName: callFrame.functionName,
+        lineNumber: callFrame.lineNumber,
+        url: callFrame.url,
+      };
+    });
+
+    const bidiStackTrace = stackFrames
+      ? {
+          callFrames: stackFrames,
+        }
+      : undefined;
+
+    return bidiStackTrace;
+  }
+
+  private _initializeLogEntryAddedEventListener() {
+    this._cdpClient.Runtime.on('consoleAPICalled', async (params) => {
+      const args = await Promise.all(
+        params.args.map(async (arg) => {
+          return await this._serializeCdpObject(arg);
+        })
+      );
+
+      await this._bidiServer.sendMessage({
+        method: 'log.entryAdded',
+        params: {
+          level: this._getLogLevel(params.type),
+          text: this._getRemoteValuesText(args, true),
+          timestamp: params.timestamp,
+          stackTrace: this._getBidiStackTrace(params.stackTrace),
+          type: 'console',
+          method: params.type,
+          realm: this._contextId,
+          args: args,
+        },
+      });
+    });
+  }
+
+  // TODO sadym: `dummyContextObject` needed for the running context.
+  // Use the proper `executionContextId` instead:
+  // https://github.com/GoogleChromeLabs/chromium-bidi/issues/52
+  private async _getDummyContextId(): Promise<string> {
+    const dummyContextObject = await this._cdpClient.Runtime.evaluate({
+      expression: '(()=>{return {}})()',
+    });
+    return dummyContextObject.result.objectId!;
+  }
+
+  /**
+   * Serializes a given CDP object into BiDi, keeping references in the
+   * target's `globalThis`.
+   * @param cdpObject CDP remote object to be serialized.
+   */
+  private async _serializeCdpObject(
+    cdpObject: Protocol.Runtime.RemoteObject
+  ): Promise<CommonDataTypes.RemoteValue> {
+    const response = await this._cdpClient.Runtime.callFunctionOn({
+      functionDeclaration: `${this.EVALUATOR_SCRIPT}.serialize`,
```

Comment:
The [`${this.EVALUATOR_SCRIPT}.serialize`](https://github.com/GoogleChromeLabs/chromium-bidi/blob/59c7c4c61ffd41a38dcc668bbdd79f224a938cf5/src/bidiMapper/scripts/eval.es#L34) method accepts the second argument `maxDepth` which is `1` by default. You can use `maxDepth=2` and serialize all the arguments in one go by putting them into an array instead of making a call per argument.